### PR TITLE
rgw/notifications: send correct size in case of delete marker creation

### PR DESCRIPTION
**Pull request by yuvalif**
Fixes: https://tracker.ceph.com/issues/51681

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/radosgw/s3-notification-compatibility.rst [hunk 1]
```diff
@@ -105,6 +105,10 @@ Event Types
 | ``s3:ReducedRedundancyLostObject``           | Not applicable to Ceph                                      |
 +----------------------------------------------+-----------------+-------------------------------------------+
 
+.. note:: 
+
+   The ``s3:ObjectRemoved:DeleteMarkerCreated`` event present information on the latest version of the object
```
Comment (mattbenjamin) [inline doc/radosgw/s3-notification-compatibility.rst:110]: presents?

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -4821,7 +4824,8 @@ void RGWDeleteObj::execute(optional_yield y)
         }
       }
     } else {
-      attrs = s->object->get_attrs();
+      attrs = std::move(astate->attrset);
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:4827]: nice catch

File: src/test/rgw/bucket_notification/test_bn.py [hunk 3]
```diff
@@ -2006,26 +2009,30 @@ def test_ps_s3_versioned_deletion_on_master():
...
                 assert False 
             else:
                 print('version ok: '+version+' in: '+str(versions))
             if event['eventName'] == 'ObjectRemoved:Delete':
                 delete_events += 1
+                assert size in [size1, size2]
                 assert event['s3']['configurationId'] in [notification_name+'_1', notification_name+'_3']
             if event['eventName'] == 'ObjectRemoved:DeleteMarkerCreated':
                 delete_marker_create_events += 1
+                assert size == size2
```
Comment (mattbenjamin) [inline src/test/rgw/bucket_notification/test_bn.py:2024]: ok

**Review by mattbenjamin — APPROVED** [review]
lgtm

File: src/rgw/rgw_op.cc (hunk 2 above)
Comment (cbodley) [inline src/rgw/rgw_op.cc:4827]: i don't think we should mutate this `RGWObjState` that's cached in `RGWObjectCtx` here; we'll read it again later in the op

**Comment by yuvalif** [comment]
jenkins render docs

File: src/rgw/rgw_op.cc (hunk 2 above)
Comment (yuvalif) [reply src/rgw/rgw_op.cc:4827]: ok. but to avoid copying, can i cache a pointer to the attribute set?

Comment (cbodley) [reply src/rgw/rgw_op.cc:4827]: as long as you don't access that pointer after a potential invalidate.. `del_op->delete_obj(this, y);` could invalidate, and maybe `s->object->swift_versioning_restore(s->obj_ctx, ver_restored, this);` as well

**Comment by yuvalif** [comment]
teuthology results: http://pulpito.front.sepia.ceph.com/yuvalif-2021-07-17_09:16:24-rgw-wip-yuval-testing1-distro-basic-smithi/
15 failures:
* bucket notification tests failed - seems like a test issue since the same failures are seen on a branch with no changes related to notifications: [here](http://qa-proxy.ceph.com/teuthology/yuvalif-2021-07-17_08:42:14-rgw-wip-yuval-testing-distro-basic-smithi/6277365/teuthology.log)
* `too many values to unpack (expected 1)` - 6 tests
* `1 pg degraded` - 3 tests
* 3 sse kms tests failed
* crash in s3 test. same crash is seen on a different branch: [here](http://qa-proxy.ceph.com/teuthology/yuvalif-2021-07-17_08:42:14-rgw-wip-yuval-testing-distro-basic-smithi/6277365/teuthology.log)
```
ceph version 17.0.0-6108-g0c4e6b6e (0c4e6b6e12f935fc82bd4e6e52ffbbcd82c66d82) quincy (dev)
 1: /lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0) [0x7f31bb07f3c0]
 2: /lib/x86_64-linux-gnu/libc.so.6(+0x18e885) [0x7f31bae0a885]
 3: (StackStringBuf<4096ul>::xsputn(char const*, long)+0x298) [0x7f31bb5330d8]
 4: (std::basic_ostream<char, std::char_traits<char> >& std::__ostream_insert<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*, long)+0x154) [0x7f31bafbc9b4]
 5: (RGWHandler_REST_S3Website::retarget(RGWOp*, RGWOp**, optional_yield)+0x404) [0x7f31bbb32c14]
 6: (rgw_process_authenticated(RGWHandler_REST*, RGWOp*&, RGWRequest*, req_state*, optional_yield, bool)+0x1a80) [0x7f31bb626d50]
 7: (process_request(rgw::sal::Store*, RGWREST*, RGWRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw::auth::StrategyRegistry const&, RGWRestfulIO*, OpsLogSocket*, optional_yield, rgw::dmclock::Scheduler*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >*, int*)+0x3b21) [0x7f31bb62be41]
 8: /lib/libradosgw.so.2(+0x4ba36d) [0x7f31bb54736d]
 9: /lib/libradosgw.so.2(+0x4bb751) [0x7f31bb548751]
 10: /lib/libradosgw.so.2(+0x4bb8cc) [0x7f31bb5488cc]
 11: make_fcontext()
```

**Comment by yuvalif** [comment]
> teuthology results: http://pulpito.front.sepia.ceph.com/yuvalif-2021-07-17_09:16:24-rgw-wip-yuval-testing1-distro-basic-smithi/
> 15 failures:
> 
> * bucket notification tests failed - seems like a test issue since the same failures are seen on a branch with no changes related to notifications: [here](http://qa-proxy.ceph.com/teuthology/yuvalif-2021-07-17_08:42:14-rgw-wip-yuval-testing-distro-basic-smithi/6277365/teuthology.log)
> * `too many values to unpack (expected 1)` - 6 tests
> * `1 pg degraded` - 3 tests
> * 3 sse kms tests failed
> * crash in s3 test. same crash is seen on a different branch: [here](http://qa-proxy.ceph.com/teuthology/yuvalif-2021-07-17_08:42:14-rgw-wip-yuval-testing-distro-basic-smithi/6277365/teuthology.log)
> 
> ```
> ceph version 17.0.0-6108-g0c4e6b6e (0c4e6b6e12f935fc82bd4e6e52ffbbcd82c66d82) quincy (dev)
>  1: /lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0) [0x7f31bb07f3c0]
>  2: /lib/x86_64-linux-gnu/libc.so.6(+0x18e885) [0x7f31bae0a885]
>  3: (StackStringBuf<4096ul>::xsputn(char const*, long)+0x298) [0x7f31bb5330d8]
>  4: (std::basic_ostream<char, std::char_traits<char> >& std::__ostream_insert<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*, long)+0x154) [0x7f31bafbc9b4]
>  5: (RGWHandler_REST_S3Website::retarget(RGWOp*, RGWOp**, optional_yield)+0x404) [0x7f31bbb32c14]
>  6: (rgw_process_authenticated(RGWHandler_REST*, RGWOp*&, RGWRequest*, req_state*, optional_yield, bool)+0x1a80) [0x7f31bb626d50]
>  7: (process_request(rgw::sal::Store*, RGWREST*, RGWRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw::auth::StrategyRegistry const&, RGWRestfulIO*, OpsLogSocket*, optional_yield, rgw::dmclock::Scheduler*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >*, int*)+0x3b21) [0x7f31bb62be41]
>  8: /lib/libradosgw.so.2(+0x4ba36d) [0x7f31bb54736d]
>  9: /lib/libradosgw.so.2(+0x4bb751) [0x7f31bb548751]
>  10: /lib/libradosgw.so.2(+0x4bb8cc) [0x7f31bb5488cc]
>  11: make_fcontext()
> ```

failures are known issues.
bucket notification tests are now passing: http://pulpito.front.sepia.ceph.com/yuvalif-2021-07-19_14:52:02-rgw:notifications-wip-yuval-testing1-distro-basic-smithi/

**Review by cbodley — APPROVED** [review]

**Pull request merged**